### PR TITLE
Add 'Open Build Scan' to the tasks and executions view

### PR DESCRIPTION
**Pull request by donat**

### CODE REVIEW

**Comment by donat** [comment]
The PR is not 100% complete, there are two things missing:
- Integration tests with different build scan plugin versions
- The icon for the toolbar
I'm thinking about creating a UI test where a project with Gradle version X and build scans version Y is imported and a task is executed. Is it OK to create a bunch of build scans just for the sake of testing?
/cc @etiennestuder

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/scan/internal/DefaultBuildScanRegistry.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.eclipse.buildship.core.scan.BuildScanRegistry;
+
+/**
+ * Default implementation for {@link BuildScanRegistry}.
+ *
+ * @author Donat Csikos
+ */
+public final class DefaultBuildScanRegistry implements BuildScanRegistry {
+
+    private final Map<ProcessDescription, String> buildScans = Collections.synchronizedMap(new HashMap<ProcessDescription, String>());
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/scan/internal/DefaultBuildScanRegistry.java]: We can't have a service that grows indefinitely. The information about the last build scan should instead be stored somewhere in the execution page.

Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/scan/internal/DefaultBuildScanRegistry.java]: Fixed.

**Comment by donat** [comment]
I've added the icon for the action. Personally, I like the artwork, but please don't hesitate to raise any concerns in that regard.

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/OpenBuildScanActionTest.groovy [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+        then:
+        view.pages.size() == 2
+        view.pages[0].openBuildScanAction.enabled
+        view.pages[0].openBuildScanAction.buildScanUrl == 'https://scans.gradle.com/s/A'
+        view.pages[1].openBuildScanAction.enabled
+        view.pages[1].openBuildScanAction.buildScanUrl == 'https://scans.gradle.com/s/B'
+    }
+
+    @Ignore
```
Comment (oehme) [inline org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/OpenBuildScanActionTest.groovy]: Needs to be un-ignored, can use our dogfooding instance.

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/OpenBuildScanActionTest.groovy [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        new RunGradleBuildLaunchRequestJob(createLaunch(task, projectDir, arguments)).schedule()
+        waitForGradleJobsToFinish()
+        waitForBuildFinishedMessage()
+    }
+
+    private void waitForBuildFinishedMessage() {
+        waitFor { consoleListener.activeConsole.document.get().contains('BUILD SUCCESSFUL') }
+    }
+
+    private void waitFor(int timeout = 5000, Closure condition) {
```
Comment (oehme) [inline org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/OpenBuildScanActionTest.groovy]: Is there no better way to listen to the console?

Comment (donat) [inline org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/OpenBuildScanActionTest.groovy]: Unfortunately not. The console output changes asynchronously, so the build job finishes before the build success message is printed.

**Comment by donat** [comment]
I added a few unit tests what in fact showed that the pattern is not good enough. After searching the interwebs I've found that there's a nice Android utility class that contains regexps for web URLs, which I copied over.

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/OpenBuildScanActionTest.groovy (hunk 3 above)
Comment (donat) [inline org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/OpenBuildScanActionTest.groovy]: I found a little improvement for the waitFor condition, see the latest changes.

File: org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/OpenBuildScanActionTest.groovy (hunk 2 above)
Comment (donat) [inline org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/execution/OpenBuildScanActionTest.groovy]: Fixed.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/console/BuildScanPatternMatchListener.java [hunk 4]
```diff
@@ -56,7 +57,7 @@ public void matchFound(PatternMatchEvent event) {
 
     @Override
     public String getPattern() {
-        return "Publishing build information...\\s+https?://[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]";
+        return "Publishing build information...\\s+" + PatternUtils.WEB_URL_PATTERN;
```
Comment (oehme) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/console/BuildScanPatternMatchListener.java]: I'd probably not match on the three dots

Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/console/BuildScanPatternMatchListener.java]: I replaced `...` with `.*` in the pattern.

**Pull request merged**